src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp

```diff
@@ -82,7 +86,142 @@ namespace AppInstaller::CLI::Workflow
+
+        if (packageVersion)
+        {
+            // TODO open composite source: package + installed
+            // how to execute without progress?
+            //auto installedSource = context.Reporter.ExecuteWithProgress(std::bind(Repository::OpenPredefinedSource, PredefinedSource::Installed, std::placeholders::_1), true);
+            //auto compositeSource = Repository::CreateCompositeSource(installedSource, packageVersion->GetSource());
+            auto compositeSource = packageVersion->GetSource();
+            context.Add<Execution::Data::DependencySource>(compositeSource);
+        }
+        else
+        {
+            // TODO question to John: can/should we do nothing for local manifests? or set up something like --dependency-source
+            //const auto& manifest = context.Get<Execution::Data::Manifest>();
+        }
+    }
+
+    void BuildPackageDependenciesGraph(Execution::Context& context)
+    {
+        context << OpenDependencySource;
+        const auto& source = context.Get<Execution::Data::DependencySource>();
+
+        std::vector<Dependency> toCheck;
+
+        // should this dictionary have Dependency as key? (to have min version) in this case Dependency should implement equal
+        std::map<string_t, std::vector<Dependency>> dependencyGraph; // < package Id, dependencies > value should be a set instead of a vector?
+
+        const auto& installer = context.Get<Execution::Data::Installer>();
+        if (installer)
+        {
+            dependencyGraph[installer->ProductId] = std::vector<Dependency>(); // ProductId is the same Id as the one used by Dependencies?
+            installer->Dependencies.ApplyToType(DependencyType::Package, [&](Dependency dependency)
+                {
+                    toCheck.push_back(dependency);
+                    dependencyGraph[dependency.Id] = std::vector<Dependency>();
+                    dependencyGraph[installer->ProductId].push_back(dependency);
+
+                });
+        } // TODO fail otherwise
+
+        for (const auto& dependencyNode : toCheck)
+        {
+            // search the package source+installed to see if the dep exists
+            SearchRequest searchRequest;
+            searchRequest.Filters.emplace_back(PackageMatchFilter(PackageMatchField::Id, MatchType::CaseInsensitive, dependencyNode.Id));
+            const auto& matches = source->Search(searchRequest).Matches;
+
+            if (!matches.empty())
+            {
+                const auto& match = matches.at(0); // What to do if there's more than one? TODO should not happen (report to the user)
+                const auto& package = match.Package;
+                if (!package->GetInstalledVersion())
+                {
+                    const auto& packageVersion = package->GetLatestAvailableVersion(); //TODO check availability before using
+                    //how to choose best installer, should I use SelectInstaller or not?

```

Comment:
I would refactor the code to make the existing task call a helper function and leverage both here.  The difference is that the workflow task is more about presentation while the helper is doing the actual selection.

Reply:
I think a refactor will not be needed, I can use ManifestComparator to get the PreferredInstaller (as used in SelectInstaller)

Reply:
You can, but I'm concerned that if you don't use common code things will diverge.  Since we pull things out of context to make the decision, we want that to happen for the dependencies as well (I believe).

Reply:
ok